cmd/exchaind/fss/check.go

```diff
@@ -0,0 +1,101 @@
+package fss
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"path/filepath"
+
+	"github.com/okex/exchain/app/utils/appstatus"
+	"github.com/okex/exchain/cmd/exchaind/base"
+	sdk "github.com/okex/exchain/libs/cosmos-sdk/types"
+	"github.com/okex/exchain/libs/iavl"
+	dbm "github.com/okex/exchain/libs/tm-db"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+func init() {
+	fssCmd.AddCommand(checkCmd)
+}
+
+// checkCmd represents the create command
+var checkCmd = &cobra.Command{
+	Use:   "check",
+	Short: "Check with the fast index with IAVL original nodes",
+	Long: `Check fast index with IAVL original nodes:
+This command is a tool to check the IAVL fast index.`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		storeKeys := appstatus.GetAllStoreKeys()
+		return check(storeKeys)
+	},
+}
+
+func check(storeKeys []string) error {
+	if !appstatus.IsFastStorageStrategy() {
+		return fmt.Errorf("db haven't upgraded to fast IAVL")
+	}
+
+	dataDir := viper.GetString(flagDataDir)
+	dbBackend := viper.GetString(sdk.FlagDBBackend)
+	db, err := base.OpenDB(filepath.Join(dataDir, base.AppDBName), dbm.BackendType(dbBackend))
+	if err != nil {
+		return fmt.Errorf("error opening dir %v backend %v DB: %w", dataDir, dbBackend, err)
+	}
+	defer db.Close()
+
+	for _, key := range storeKeys {
+		prefix := []byte(fmt.Sprintf("s/k:%s/", key))
+		prefixDB := dbm.NewPrefixDB(db, prefix)
+		log.Printf("Checking.... %v\n", key)
+
+		mutableTree, err := iavl.NewMutableTree(prefixDB, 0)
+		if err != nil {
+			return err
+		}
+		if _, err := mutableTree.LoadVersion(0); err != nil {
+			return err
+		}
+
+		if err := checkIndex(mutableTree); err != nil {
+			return fmt.Errorf("%v iavl fast index not match %v", key, err.Error())
+		}
+	}
+	log.Println("Success")
+
+	return nil
+}
+
+func checkIndex(mutableTree *iavl.MutableTree) error {
+	fastIterator := mutableTree.Iterator(nil, nil, true)
+	defer fastIterator.Close()
+	iterator := iavl.NewIterator(nil, nil, true, mutableTree.ImmutableTree)
```

Comment:
If the iterator all keys in once NewIterator,  the memory should larage?

Reply:
The memory issue is not significant, there is no oom yet, because we only read the nodes. but when we create fast-storage, the memory issue is significant, it will panic because the batch's size. so I think here is ok.